Catch errors when publishing private packages

https://fusion2.corp.google.com/invocations/1c0263ab-ffe9-4202-8c2b-b48ab014215a/targets/cloud-devrel%2Fclient-libraries%2Fnodejs%2Frelease%2Fgoogleapis%2Fgoogle-api-nodejs-client%2Fpublish/log

This publish fails when attempting to publish samples/package.json. The fix will be to add `private: true`, like so: https://github.com/googleapis/google-api-nodejs-client/pull/3139

Npm will fail when attempting to publish, so we need to catch errors when it attempts to publish private packages. Example output below:

```
npm notice Publishing to https://registry.npmjs.org/
npm ERR! code EPRIVATE
npm ERR! This package has been marked as private
npm ERR! Remove the 'private' field from the package.json to publish it.
{
  "error": {
    "code": "EPRIVATE",
    "summary": "This package has been marked as private\nRemove the 'private' field from the package.json to publish it.",
    "detail": ""
  }
}
```

--- a/packages/mono-repo-publish/src/main.ts
+++ b/packages/mono-repo-publish/src/main.ts
@@ -122,10 +122,14 @@ function publish(
   execSync: typeof childProcess.execSync,
   rmSync: typeof fs.rmSync
 ): ExecutionOutput {
-  const pkg = fs.readFileSync(resolve(directory, 'package.json'))
-  if (pkg.private) {
-    return {
-      output: "skipping publication ${directory}/package.json is private"
+  if (fs.existsSync(resolve(directory, 'package.json'))) {
+    const pkg = JSON.parse(
+      fs.readFileSync(resolve(directory, 'package.json'), 'utf-8')
+    );
+    if (pkg.private) {
+      return {
+        output: 'skipping publication ${directory}/package.json is private',
+      };
     }
   }
   const installCommand = stat(resolve(directory, 'package-lock.json'))

--- a/packages/mono-repo-publish/src/main.ts
+++ b/packages/mono-repo-publish/src/main.ts
@@ -122,7 +122,7 @@ function publish(
   execSync: typeof childProcess.execSync,
   rmSync: typeof fs.rmSync
 ): ExecutionOutput {
-  const pkg = readFileSync(resolve(directory, 'package.json'))
+  const pkg = fs.readFileSync(resolve(directory, 'package.json'))
   if (pkg.private) {
     return {
       output: "skipping publication ${directory}/package.json is private"

--- a/packages/mono-repo-publish/src/main.ts
+++ b/packages/mono-repo-publish/src/main.ts
@@ -90,7 +90,6 @@ export function listChangedSubmodules(
   excludeGlobs: string[] = []
 ): string[] {
   const globs = excludeGlobs.map(glob => new mm.Minimatch(glob));
-  // Only checking for package.jsons in submodules that were changed
   // Not checking the top-level package.json
   let files = prFiles.filter(file =>
     file.match(/\/package\.json$|^package\.json$/)
@@ -123,6 +122,12 @@ function publish(
   execSync: typeof childProcess.execSync,
   rmSync: typeof fs.rmSync
 ): ExecutionOutput {
+  const pkg = readFileSync(resolve(directory, 'package.json'))
+  if (pkg.private) {
+    return {
+      output: "skipping publication ${directory}/package.json is private"
+    }
+  }
   const installCommand = stat(resolve(directory, 'package-lock.json'))
     ? 'ci'
     : 'i';

--- a/packages/mono-repo-publish/test/main.test.ts
+++ b/packages/mono-repo-publish/test/main.test.ts
@@ -160,4 +160,13 @@ describe('mono-repo publish', () => {
       'npm publish --access=public'
     );
   });
+
+  it('skips package if package is set to private', () => {
+    const output = core.publishSubmodules(
+      ['test/sample-private-package'],
+      false
+    );
+    console.log(output);
+    assert.deepStrictEqual(output, []);
+  });
 });

--- a/packages/mono-repo-publish/test/sample-private-package/package.json
+++ b/packages/mono-repo-publish/test/sample-private-package/package.json
@@ -1,0 +1,36 @@
+{
+    "name": "googleapis-samples",
+    "description": "The following samples show basic usage of various APIs.",
+    "repository": "google/google-api-nodejs-client",
+    "license": "Apache-2.0",
+    "private": true,
+    "engines": {
+      "node": ">=12.0.0"
+    },
+    "files": [
+      "**/*.js",
+      "!test/"
+    ],
+    "scripts": {
+      "prepare": "cd typescript && tsc && cd ..",
+      "test": "mocha"
+    },
+    "dependencies": {
+      "@google-cloud/local-auth": "^2.0.0",
+      "express": "^4.17.1",
+      "googleapis": "^112.0.0",
+      "googleapis-common": "^6.0.0",
+      "nconf": "^0.12.0",
+      "open": "^8.0.0",
+      "server-destroy": "^1.0.1",
+      "uuid": "^9.0.0"
+    },
+    "devDependencies": {
+      "chai": "^4.2.0",
+      "execa": "^5.0.0",
+      "mocha": "^9.0.0",
+      "nock": "^13.0.0",
+      "proxyquire": "^2.1.3",
+      "typescript": "^4.6.4"
+    }
+  }

